// @progress/kendo-react-upload/dist/npm/messages/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var _a;
/**
 * @hidden
 */
exports.cancel = 'upload.cancel';
/**
 * @hidden
 */
exports.clearSelectedFiles = 'upload.clearSelectedFiles';
/**
 * @hidden
 */
exports.dropFilesHere = 'upload.dropFilesHere';
/**
 * @hidden
 */
exports.headerStatusUploaded = 'upload.headerStatusUploaded';
/**
 * @hidden
 */
exports.headerStatusUploading = 'upload.headerStatusUploading';
/**
 * @hidden
 */
exports.invalidFileExtension = 'upload.invalidFileExtension';
/**
 * @hidden
 */
exports.invalidFiles = 'upload.invalidFiles';
/**
 * @hidden
 */
exports.invalidMaxFileSize = 'upload.invalidMaxFileSize';
/**
 * @hidden
 */
exports.invalidMinFileSize = 'upload.invalidMinFileSize';
/**
 * @hidden
 */
exports.remove = 'upload.remove';
/**
 * @hidden
 */
exports.retry = 'upload.retry';
/**
 * @hidden
 */
exports.select = 'upload.select';
/**
 * @hidden
 */
exports.uploadSelectedFiles = 'upload.uploadSelectedFiles';
/**
 * @hidden
 */
exports.total = 'upload.total';
/**
 * @hidden
 */
exports.files = 'upload.files';
/**
 * @hidden
 */
exports.statusUploaded = 'upload.statusUploaded';
/**
 * @hidden
 */
exports.statusUploadFailed = 'upload.statusUploadFailed';
/**
 * @hidden
 */
exports.dropZoneHint = 'upload.dropZoneHint';
/**
 * @hidden
 */
exports.dropZoneNote = 'upload.dropZoneNote';
/**
 * @hidden
 */
exports.messages = (_a = {},
    _a[exports.cancel] = 'Cancel',
    _a[exports.clearSelectedFiles] = 'Clear',
    _a[exports.dropFilesHere] = 'Drop files here to upload',
    _a[exports.headerStatusUploaded] = 'Done',
    _a[exports.headerStatusUploading] = 'Uploading...',
    _a[exports.invalidFileExtension] = 'File type not allowed.',
    _a[exports.invalidFiles] = 'Invalid file(s). Please check file upload requirements.',
    _a[exports.invalidMaxFileSize] = 'File size too large.',
    _a[exports.invalidMinFileSize] = 'File size too small.',
    _a[exports.remove] = 'Remove',
    _a[exports.retry] = 'Retry',
    _a[exports.select] = 'Select files...',
    _a[exports.uploadSelectedFiles] = 'Upload',
    _a[exports.total] = 'Total',
    _a[exports.files] = 'files',
    _a[exports.statusUploaded] = 'File(s) successfully uploaded.',
    _a[exports.statusUploadFailed] = 'File(s) failed to upload.',
    _a[exports.dropZoneHint] = 'Drag and drop files here to upload.',
    _a[exports.dropZoneNote] = 'Only JPEG and PNG files are allowed.',
    _a);
